File: core-golang/notifier/email.go
package notifier

import (
	"fmt"
	"net/smtp"
)

// sendEmail sends a plain-text email notification via SMTP with PlainAuth.
func sendEmail(cfg EmailConfig, event ChangeEvent) error {
	subject := fmt.Sprintf("Vigil — Change detected on %s", event.SiteID)
	body := fmt.Sprintf(
		"Change detected by Vigil\n\n"+
			"Site: %s\n"+
			"URL: %s\n"+
			"Detected: %s\n\n"+
			"Change summary:\n%s\n",
		event.SiteID,
		event.URL,
		event.DetectedAt,
		event.DiffSummary,
	)

	msg := fmt.Sprintf(
		"From: %s\r\n"+
			"To: %s\r\n"+
			"Subject: %s\r\n"+
			"Content-Type: text/plain; charset=UTF-8\r\n"+
			"\r\n"+
			"%s",
		cfg.SMTPFrom,
		cfg.NotifyEmailTo,
		subject,
		body,
	)

	addr := fmt.Sprintf("%s:%s", cfg.SMTPHost, cfg.SMTPPort)
	auth := smtp.PlainAuth("", cfg.SMTPFrom, cfg.SMTPPassword, cfg.SMTPHost)

	err := smtp.SendMail(addr, auth, cfg.SMTPFrom, []string{cfg.NotifyEmailTo}, []byte(msg))
	if err != nil {
		return fmt.Errorf("email: send: %w", err)
	}
	return nil
}

File: core/notifier/email.ts
import * as nodemailer from "nodemailer";
import type { Event } from "./notifier";

export interface EmailConfig {
    host: string;
    port: string;
    from: string;
    password?: string;
    to: string;
}

export function loadEmailConfig(): EmailConfig {
    return {
        host: process.env.SMTP_HOST || "",
        port: process.env.SMTP_PORT || "",
        from: process.env.SMTP_FROM || "",
        password: process.env.SMTP_PASSWORD || "",
        to: process.env.NOTIFY_EMAIL_TO || "",
    };
}

export function createTransporter(config: EmailConfig) {
    return nodemailer.createTransport({
        host: config.host,
        port: parseInt(config.port) || 587,
        secure: parseInt(config.port) === 465,
        auth: config.password ? { user: config.from, pass: config.password } : undefined,
    });
}

export async function sendEmail(
    config: EmailConfig,
    transporter: nodemailer.Transporter,
    event: Event,
) {
    await transporter.sendMail({
        from: config.from,
        to: config.to,
        subject: `🚨 Vigil Alert: ${event.url}`,
        text: `Site: ${event.url}\nDiff: ${event.diff_summary}\nDetected: ${event.detected_at}`,
        html: `<p>🚨 <b>Vigil Alert</b></p><p><b>Site:</b> ${event.url}</p><p><b>Diff:</b> ${event.diff_summary}</p><p><b>Detected:</b> ${event.detected_at}</p>`,
    });
}
